### PR TITLE
Make controller-manager resilient to stale serviceaccount tokens

### PR DESCRIPTION
**Pull request by liggitt**
Now that the controller manager is spinning up controller loops using service accounts, we need to be more proactive in making sure the clients will actually work.

Future additional work:
* make a controller that reaps invalid service account tokens (c.f. https://github.com/kubernetes/kubernetes/issues/20165)
* allow updating the client held by a controller with a new token while the controller is running (c.f. https://github.com/kubernetes/kubernetes/issues/4672)

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-auth-pr-reviews

**Comment by liggitt** [comment]
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this

File: pkg/controller/client_builder.go [hunk 1]
```diff
@@ -180,6 +188,69 @@ func (b SAControllerClientBuilder) Config(name string) (*restclient.Config, erro
...
+		if tokenResult.Status.User.Username != username {
+			glog.Warningf("Token for %s/%s authenticated as unexpected username: %s", sa.Name, sa.Namespace, tokenResult.Status.User.Username)
+			return nil, false, nil
+		}
+		glog.V(4).Infof("Verified credential for %s/%s", sa.Name, sa.Namespace)
+		return clientConfig, true, nil
+	}
+
+	// If we couldn't run the token review, the API might be disabled or we might not have permission.
+	// Try to make a request to /apis with the token. If we get a 401 we should consider the token invalid.
```
Comment (deads2k) [inline pkg/controller/client_builder.go]: blech.  Mostly for GKE, right?  We should start requiring that API to be on.

Comment (liggitt) [inline pkg/controller/client_builder.go]: yes. @cjcullen, status on this?

Comment (deads2k) [inline pkg/controller/client_builder.go]: @pwittrock just hit this too.  Delegated authentication works on other platforms, so a aggregated apiserver pod author would end up trying to target platforms. GKE should enable the endpoint.

Comment (liggitt) [inline pkg/controller/client_builder.go]: don't disagree, but I still want the fallback here

**Pull request merged**